### PR TITLE
MERGE EXTENSION: when not matched by source - update

### LINKED ISSUE #860: Add update action for MERGE NOT MATCHED BY SOURCE
(Auth: @sabrinas-tuidi)

We are currently running our dbt project on pre-released version 1.9.0, which includes your contribution for matched and not matched conditions. This is super useful so thank you so much for working on this! 
We are currently facing a blocker since, due to the nature of our workflow, we cannot perform a direct delete action when not matched by source. Instead, we need to run an update set on a flag that indicated the rows to be deleted. 
This is an example: 
```
      MERGE INTO {catalog}.{silver_database}.purchase_promotion r
      USING tab_update u
      ON r.purchase_promotion_row_code = u.purchase_promotion_row_code  -- Match on promotion_row_code
      AND r.promotion_code = u.promotion_code
      -- If a row in the target (r) is not matched by any row in the source (u)
      WHEN NOT MATCHED BY SOURCE
      AND r.promotion_end_date > CAST('{date_from}' AS DATE)
      THEN
        UPDATE SET
          r.update_date = CAST('{update_date}' AS TIMESTAMP),
          r.to_be_deleted = 1
```
We understand only the delete action is supported at the moment. 

_Originally posted by @sabrinas-tuidi in https://github.com/databricks/dbt-databricks/issues/245#issuecomment-2512234624_

### PR DESCRIPTION
**Pull request by mi-volodin**
Resolves #860 

### Description

This PR will address the missing use case shared in the issue #860. When there's no match by source and condition holds now it will be possible to set update rule manually (will be executed syntactically as part of `update set` block).

### Checklist

- [x] I have run this code in development and it appears to resolve the stated issue
- [x] This PR includes tests, or tests are not required/relevant for this PR
- [x] I have updated the `CHANGELOG.md` and added information about my change to the "dbt-databricks next" section.

### CODE REVIEW

**Comment by mi-volodin** [comment]
@benc-db Hi, I am having fun again with the `pre-commit` with ruff.
Line length is set to 100 and this line violates `ruff` check

```python
        return {
            "not_matched_by_source_expected.csv": fixtures.not_matched_by_source_then_delete_expected,
        }
```

If fixed to multiline it violates `ruff-format` check, because it gets formatted as in previous line.
```python
        return {
            "not_matched_by_source_expected.csv":
                fixtures.not_matched_by_source_then_delete_expected,
        }
```

To avoid this I will just reduce the name sizes and trade the readability off. But JFYI

**Comment by mi-volodin** [comment]
To set up the environment I used this

```bash
uv venv
source .venv/bin/activate

uv install .
```

**Comment by benc-db** [comment]
> (**BTW** Really happy that you have switched towards ruff/astral thingy 🙏 , though I am missing the guideline to run the tests and set up the environment.)

In the doc folder, dbt-databricks-dev.  If after setting things up per those instructions you are still having ruff behave weirdly, see if #noqa after the line helps.

**Comment by mi-volodin** [comment]
@benc-db thanks, worked out for me.

For transparency - managed to run integration tests, currently experience issues with suggested test cases.

**Comment by mi-volodin** [comment]
Tests are "passing" on my side. Still have same list of issues like previous time on my cluster with Snapshots. 
That's why I haven't set-up the GH-Actions to run (it will fail anyway). 

@benc-db I am confused with the update of the `CHANGELOG.md`. The 1.9.0 stated to be (TBD) with release date, therefore I updated the changelog for same version. Please guide me on what is better to be stated (version bump, or same entry).

**Comment by benc-db** [comment]
> @benc-db I am confused with the update of the CHANGELOG.md. The 1.9.0 stated to be (TBD) with release date, therefore I updated the changelog for same version. Please guide me on what is better to be stated (version bump, or same entry).

Let me take a look at this PR today, and validate the E2E tests.  If it's good to go, we can put it in 1.9.1, along with a couple of other fixes I'm working on today.

File: CHANGELOG.md [hunk 1]
```diff
@@ -7,7 +7,7 @@
 - Add 'user_folder_for_python' behavior to switch writing python model notebooks to the user's folder ([835](https://github.com/databricks/dbt-databricks/pull/835))
 - Merge capabilities are extended ([739](https://github.com/databricks/dbt-databricks/pull/739)) to include the support for the following features (thanks @mi-volodin):
   - `with schema evolution` clause (requires Databricks Runtime 15.2 or above);
-  - `when not matched by source` clause, only for `delete` action
```
Comment (benc-db) [inline CHANGELOG.md:10]: Please move to 1.9.1; 1.9.0 was released yesterday.

**Review by benc-db — DISMISSED** [review]

File: CHANGELOG.md (hunk 1 above)
Comment (mi-volodin) [reply CHANGELOG.md:10]: Ok, will adjust it later, when tests are done.

**Comment by benc-db** [comment]
Everything looks good, just rebase off latest main / update changelog.

**Comment by mi-volodin** [comment]
@benc-db both done

**Comment by benc-db** [comment]
Thanks so much for fleshing this out.  I really appreciate it!

**Comment by mi-volodin** [comment]
Happy to help! Thanks for your support 🙏

**Review by benc-db — APPROVED** [review]

**Comment by benc-db** [comment]
Will merge when E2E tests finish.  Will roll out with 1.9.1, probably releasing on Monday.

**Pull request merged**